### ISSUE TITLE
AgentX master agent failed to respond to ping

### ISSUE THREAD

**Issue by fulljackz** [acassen/keepalived#2195]
**Describe the bug**
About every 5 days after keepalived restart (or reboot), snmpd config drop down and this log appears :

```
Keepalived_vrrp[680]: AgentX master agent failed to respond to ping.  Attempting to re-register.
Keepalived_vrrp[680]: NET-SNMP version 5.9 AgentX subagent connected
```

Then, when I try to snmpwalk the keepalived mib, I have an error : 
```
Error in packet.
Reason: (genError) A general failure occured
Failed object: SNMPv2-SMI::enterprises.9586.100
```

Randomly, something happens and everything come back again with this in logs : 
```
Sep  8 10:10:01 mysrv Keepalived_vrrp[680]: AgentX master agent failed to respond to ping.  Attempting to re-register.
Sep  8 10:10:02 mysrv Keepalived_vrrp[680]: scheduler: Error performing epoll_ctl DEL op for fd:18 (Bad file descriptor)
Sep  8 10:10:02 mysrv Keepalived_vrrp[680]: A thread timer expired 1.000899 seconds ago
Sep  8 10:10:17 mysrv Keepalived_vrrp[680]: NET-SNMP version 5.9 AgentX subagent connected
Sep  8 10:11:30 mysrv Keepalived_vrrp[680]: AgentX master disconnected us, reconnecting in 15
Sep  8 10:11:30 mysrv Keepalived_vrrp[680]: scheduler: Error performing epoll_ctl DEL op for fd:13 (No such file or directory)
Sep  8 10:11:45 mysrv Keepalived_vrrp[680]: NET-SNMP version 5.9 AgentX subagent connected
```
After that, attempting to snmpwalk is ok.

On the secondary server everything seems ok.

**To Reproduce**
Any steps necessary to reproduce the behaviour:
- start keepalived with snmp support
- snmpwalk every 5 minutes (monitoring)
- randomly keepalived drop snmp support and snmpwalk is unavailable

**Expected behavior**
snmpwalk working permanently

**Keepalived version**
```
Keepalived v2.1.5 (07/13,2020)

Copyright(C) 2001-2020 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 5.10.70
Running on Linux 5.10.0-16-amd64 #1 SMP Debian 5.10.127-1 (2022-06-30)

configure options: --build=x86_64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir=${prefix}/lib/x86_64-linux-gnu --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --with-kernel-dir=debian/ --enable-snmp --enable-sha1 --enable-snmp-rfcv2 --enable-snmp-rfcv3 --enable-dbus --enable-json --enable-bfd --enable-regex build_alias=x86_64-linux-gnu CFLAGS=-g -O2 -ffile-prefix-map=/build/keepalived-jCUlld/keepalived-2.1.5=. -fstack-protector-strong -Wformat -Werror=format-security LDFLAGS=-Wl,-z,relro CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2

Config options:  NFTABLES LVS REGEX VRRP VRRP_AUTH JSON BFD OLD_CHKSUM_COMPAT FIB_ROUTING SNMP_V3_FOR_V2 SNMP_VRRP SNMP_CHECKER SNMP_RFCV2 SNMP_RFCV3 DBUS

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF IPV6_ADVANCED_API LIBNL3 RTA_ENCAP RTA_EXPIRES RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS FRA_L3MDEV FRA_UID_RANGE RTAX_FASTOPEN_NO_COOKIE RTA_VIA FRA_OIFNAME FRA_PROTOCOL FRA_IP_PROTO FRA_SPORT_RANGE FRA_DPORT_RANGE RTA_TTL_PROPAGATE IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA NET_LINUX_IF_H_COLLISION LIBIPVS_NETLINK IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS IPVS_TUN_TYPE IPVS_TUN_CSUM IPVS_TUN_GRE VRRP_VMAC VRRP_IPVLAN IFLA_LINK_NETNSID CN_PROC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RESET_ON_FORK
```

**Distro (please complete the following information):**
 - Name: Debian
 - Version: 11
 - Architecture: x86_64

**Configuration file:**
```
global_defs {
enable_script_security
script_user root
  notification_email {
   root@mydomain.com
   }
enable_script_security
script_user root
   notification_email_from mysrv@mydomain.com
     smtp_server 10.11.12.254
     smtp_connect_timeout 30
   }
       
   vrrp_script check_sip {
       script       "/etc/keepalived/checksip.sh"
       interval 5   # check every 5 seconds
       fall 4       # require 2 failures for KO
       rise 4       # require 4 successes for OK
   }
       
   vrrp_instance lb-out{
       state BACKUP
       interface ens256
       virtual_router_id 50
       nopreempt
       advert_int 1
       smtp_alert
       authentication {
          auth_type PASS
          auth_pass MYPASS
       }
       virtual_ipaddress {
           10.11.12.250/24 brd 10.207.201.255 dev ens192 label ens192:0
       }
       track_script {
           check_sip
       }
       notify_master "/etc/keepalived/master-backup.sh MASTER"
       notify_backup "/etc/keepalived/master-backup.sh BACKUP"
       notify_fault "/etc/keepalived/master-backup.sh FAULT"
       }
```

**Additional context**

```
# /etc/default/keepalived
# Options to pass to keepalived

# DAEMON_ARGS are appended to the keepalived command-line
DAEMON_ARGS="-x"
```
- Restarting snmpd does not solve the issue

**Comment by pqarmitage** [comment]
This looks like a problem with snmpd. The error message `Failed object: SNMPv2-SMI::enterprises.9586.100` suggests to me that the keepalived MIB files may not be properly installed. Certainly snmpd is having a problem and stops responding, hence the message `AgentX master agent failed to respond to ping.` The error message `scheduler: Error performing epoll_ctl DEL op for fd:18` and likewise for fd 13 are probably the file descriptors used by keepalived to communicate with snmpd.

**Comment by fulljackz** [comment]
It's possible, but the reason why I created this issue is : 
- Keepalived MIB is in the same place on each of our multiple keepalived / snmpd servers
- Snmp daemon is still responding to others nagios checks 

Today the issue came back on one of the server where it disappeared automagically few days ago.
I don't really known what / where search any clues. If I `systemctl restart keepalived` it will be ok for some days again.

**Comment by pqarmitage** [comment]
`AgentX master agent failed to respond to ping.  Attempting to re-register.` is an error message returned by libnetsnmp, and it means that the snmp library failed to communicate with snmpd. The requirement for the snmp library to ping snmpd is a problem if snmpd goes away and can cause keepalived timeouts, since the keepalived process that calls the snmp library to ping snmpd (and eventually times out) is blocked until the ping is responded to or the request times out (see keepalived/core/snmp.c for more info). This blocking of the keepalived_vrrp process in the snmp library is what has caused the log message `A thread timer expired 1.000899 seconds ago`, which means that keepalived when it next ran knew that a timer expired 1.000899 seconds earlier, but it hadn't been able to process the timer in a timely fashion. 

keepalived uses epoll_wait() to multiplex across the various file descriptors that it uses. The snmp library also uses file descriptors to communicate with snmpd. The snmp library has an interface to return the file descriptors that it is using, so that the application using the library can run its own I/O scheduler, but unfortunately libnetsnmp only provides direct support for using select(), and so keepalived provides function snmp_epoll_info() to convert from what snmp_select_info() returns to what is needed to use epoll_wait(). From past experience I have discovered that the file descriptors returned by snmp_select_info() do not remain constant (presumably because libnetsnmp has a failure to communicate with snmpd has so has to open new file descriptors), and so when keepalived has a timeout returned to it by libnetsnmp it has to call snmp_select_info() via snmp_epoll_info() again in order to update the file descriptors used. The message `scheduler: Error performing epoll_ctl DEL op for fd:18 (Bad file descriptor)` relates to precisely that (i.e. libnetsnmp has changed the file descriptor that it is using, and while keepalived is attempting to delete the old file descriptor before adding the new one epoll_ctl() reports that the old file descriptors is not longer valid, presumably because libnetsnmp has closed it).

So certainly what you are experiencing is a failure in the communication between libnetsnmp and snmpd. I have discovered in the past that this _can_ be triggered by missing MIB files, but there are other reasons for it as well. Reasons for snmpd disconnecting from libnetsnmp instances are varied, and it can easily be triggered by another process doing something that snmpd doesn't like.

I don't claim to be an expert on using net-snmp, but I certainly find it horrible to use.  If there is a better way to interface to it I would happily do so, but I am not aware of any such way.

In order to track down the issue, you will need to look at the snmpd system logs. If there isn't enough information being provided, then you will need to increase the logging level of snmpd. The man page for snmpd says see LOGGING_OPTIONS in snmpcmd(1) man page, which helpfully may or may not be installed on your system.
